h/api/auth.py

```diff
@@ -19,3 +22,96 @@ def translate_annotation_principals(principals):
         else:
             result.add(principal)
     return list(result)
+
+
+def generate_bearer_token(request, expires_in):
+    """Return a signed JSON Web Token for the given request.
+
+    The token can be used in the Authorization header in subsequent requests to
+    the API to authenticate the user identified by the
+    request.authenticated_userid of the _current_ request.
+
+    :param request: the HTTP request to return a token for, the token will
+        authenticate the userid given by this request's authenticated_userid
+        property
+    :type request: pyramid.request.Request
+
+    :param expires_in: when the returned token should expire, in seconds from
+        the current time
+    :type expires_in: int
+
+    :returns: a signed JSON Web Token
+    :rtype: string
+
+    """
+    now = datetime.datetime.utcnow().replace(microsecond=0)
+
+    claims = {
+        'iss': request.registry.settings['h.client_id'],
+        'aud': request.host_url,
+        'sub': request.authenticated_userid,
+        'exp': now + datetime.timedelta(seconds=expires_in),
+        'iat': now,
+    }
+
+    return jwt.encode(claims,
+                      request.registry.settings['h.client_secret'],
+                      algorithm='HS256')
+
+
+def userid_from_bearer_token(request):
+    """Return the userid authenticated by the given request's JWT bearer token.
+
+    If the request contains a token generated by generate_bearer_token() above
+    in its Authorization header then return the userid authenticated by the
+    token.
+
+    If the request contains no token, the token is invalid, or the token
+    doesn't contain a userid, return None.
+
+    :param request: the request whose Authorization header to check
+    :type request: pyramid.request.Request
+
+    :returns: the userid authenticated by the request's JWT or None
+    :rtype: unicode or None
+
+    """
+    if 'Authorization' not in request.headers:
+        return None
+
+    try:
+        return jwt.decode(request.headers.get('Authorization')[7:],
+                          key=request.registry.settings['h.client_secret'],
+                          audience=request.host_url,
+                          leeway=240,
+                          algorithm='HS256').get('sub')
+    except jwt.InvalidTokenError:
+        return None
+
+
+def unauthenticated_userid_from_bearer_token(request):
```

Comment:
As mentioned earlier I'm not sure there's much point in doing this. I don't think we actually need or use `unauthenticated_userid` anywhere. I'm happy for it to delegate to `authenticated_userid` in the authentication policy.